crypto/evp/evp_fetch.c

```diff
@@ -107,35 +117,23 @@ static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
     return 0;
 }
 
-static void *construct_method(const char *algorithm_name,
-                              const OSSL_DISPATCH *fns, OSSL_PROVIDER *prov,
-                              void *data)
+static void *construct_method(const char *name, const OSSL_DISPATCH *fns,
+                              OSSL_PROVIDER *prov, void *data)
 {
     struct method_data_st *methdata = data;
-    void *method = NULL;
-    int nid = OBJ_sn2nid(algorithm_name);
-
-    if (nid == NID_undef) {
-        /* Create a new NID for that name on the fly */
-        ASN1_OBJECT tmpobj;
-
-        /* This is the same as OBJ_create() but without requiring a OID */
-        tmpobj.nid = OBJ_new_nid(1);
-        tmpobj.sn = tmpobj.ln = methdata->name;
-        tmpobj.flags = ASN1_OBJECT_FLAG_DYNAMIC;
-        tmpobj.length = 0;
-        tmpobj.data = NULL;
-
-        nid = OBJ_add_object(&tmpobj);
-    }
-
-    if (nid == NID_undef)
-        return NULL;
-
-    method = methdata->method_from_dispatch(nid, fns, prov);
-    if (method == NULL)
-        return NULL;
-    return method;
+    int legacy_nid = 0;
+
+#ifndef FIPS_MODE
+    /* TODO(3.x) get rid of the need for legacy NIDs */
+    /*
+     * FIPS module note: since internal fetches will be entirely provider
+     * based, we know that none of its code depends on legacy NIDs or any
+     * functionality that use them.
+     */
+    legacy_nid = OBJ_sn2nid(name);
+#endif
```

Comment:
So here's a point that may make this PR controversial.  I would like to propose that *within the FIPS module*, the "legacy" NIDs are (or should be) irrelevant.

In the particular case I identified where "legacy" NIDs are needed is all about ASN.1 parameters, something we won't have to deal with within the FIPS module as far as I understand.  I'll spend some time identifying others...

Reply:
... or perhaps not so controversial after all...  I just looked in #8728 and noticed that the "offending" code is excluded from the FIPS module, so...